### PR TITLE
add integration with kura

### PR DESCRIPTION
**Pull request by lihaoyucool**

### CODE REVIEW

File: packages/examples/kura/integration_with_kura.py [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+
+    if (msg.topic.split("/").__len__() == 4) and (msg.topic.split("/")[3] == "pulse-topic"): # msg.topic = "pulse-account-name/pulseid/pulse/pulse-topic"
+
+        Jsonpayload = json.loads(msg.payload)['metrics']
+        if Jsonpayload: # msg.payload = "{"sentOn":1533007387525,"metrics":{"assetName":"Store-GW-03","sensor_timestamp":1533007387525,"sensor":0}}"
+
+            #is new gateway?
+            if Jsonpayload['assetName'] in Gatewaydict.keys():
+
+                log.info("is old gateway")
```
Comment (KohliDev) [inline packages/examples/kura/integration_with_kura.py]: Change the log message to Gateway already registered.

File: packages/examples/kura/integration_with_kura.py [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+
+                        except Exception as e:
+                            log.error(
+                                'Exception while loading metric {0} for Edge System {1}'.format
+                                (key_metric, str(e)))
+
+
+            else: #new device enrolling
+
+                log.info("is new gateway")
```
Comment (KohliDev) [inline packages/examples/kura/integration_with_kura.py]: Change the log message to new Device to be registered.

File: packages/examples/kura/integration_with_kura.py [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+delay_retries_min = 600
+# Retry delay Max Value in seconds
+delay_retries_max = 1800
+
+# Lambda Function Multiplier uses the above settings for calculating retry and delay logic
+lfm = lambda x: x * no_of_edge_system_in_thousands
+retry_attempts = lfm(no_of_retries_for_connection)
+delay_retries = random.randint(lfm(delay_retries_min), lfm(delay_retries_max))
+
+lock = threading.Lock()
```
Comment (KohliDev) [inline packages/examples/kura/integration_with_kura.py]: Remove this if not used?

File: packages/examples/kura/integration_with_kura.py (hunk 1 above)
Comment (lihaoyucool) [reply packages/examples/kura/integration_with_kura.py]: Fixed.

File: packages/examples/kura/integration_with_kura.py (hunk 3 above)
Comment (lihaoyucool) [reply packages/examples/kura/integration_with_kura.py]: Fixed.

File: packages/examples/kura/integration_with_kura.py (hunk 2 above)
Comment (lihaoyucool) [reply packages/examples/kura/integration_with_kura.py]: Fixed.

**Comment by lihaoyucool** [comment]
@KohliDev I already addressed your comments so please kindly help us get this merged ASAP，and thanks a lot.

File: packages/examples/kura/README.txt [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+This directory includes:
+File 1: integration_with_kura.py
+
+
+How to use it:
+
+1.sha1sum integration_with_kura.py --> 16013ef7a95f59537bc5ab1d91a5c599c9cb8ab4
+2.liotapkg.sh load integration_with_kura 16013ef7a95f59537bc5ab1d91a5c599c9cb8ab4
+3.config Kura UI to publish data
```
Comment (KohliDev) [inline packages/examples/kura/README.txt]: Provide a link/step on how to configure Kura UI.

File: packages/examples/kura/integration_with_kura.py [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+
+callbackFunc_str = '''def metric_callback_%s():    return Msg_dic['%s']['msg']''' # update the data by use Msg_dic disc
+
+"""Subscibe callback function"""
+def sub_callback(self, client, userdata, msg):
+
+    if (msg.topic.split("/").__len__() == 4) and (msg.topic.split("/")[3] == "pulse-topic"): # msg.topic = "pulse-account-name/pulseid/pulse/pulse-topic"
+
+        Jsonpayload = json.loads(msg.payload)['metrics']
+        if Jsonpayload: # msg.payload = "{"sentOn":1533007387525,"metrics":{"assetName":"Store-GW-03","sensor_timestamp":1533007387525,"sensor":0}}"
```
Comment (KohliDev) [inline packages/examples/kura/integration_with_kura.py]: Provide comment above this line

File: packages/examples/kura/integration_with_kura.py [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+
+Msg_dic = {}                #{"demo": {"flag": 0, "msg": 0, "metric": metric}}
+Gatewaydict = {}            #{"assetName": reg_device}
+
+callbackFunc_str = '''def metric_callback_%s():    return Msg_dic['%s']['msg']''' # update the data by use Msg_dic disc
+
+"""Subscibe callback function"""
+def sub_callback(self, client, userdata, msg):
+
+    if (msg.topic.split("/").__len__() == 4) and (msg.topic.split("/")[3] == "pulse-topic"): # msg.topic = "pulse-account-name/pulseid/pulse/pulse-topic"
```
Comment (KohliDev) [inline packages/examples/kura/integration_with_kura.py]: provide comment above this line

File: packages/examples/kura/integration_with_kura.py [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+
+        Jsonpayload = json.loads(msg.payload)['metrics']
+        if Jsonpayload: # msg.payload = "{"sentOn":1533007387525,"metrics":{"assetName":"Store-GW-03","sensor_timestamp":1533007387525,"sensor":0}}"
+
+            #is new gateway?
```
Comment (KohliDev) [inline packages/examples/kura/integration_with_kura.py]: Add comment as:
"If new Gateway to be registered will be checked below"

File: packages/examples/kura/integration_with_kura.py [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+            if Jsonpayload['assetName'] in Gatewaydict.keys():
+
+                log.info("Device already registered.")
+
+                for key_metric in Jsonpayload.keys():
+
+                    # filter the unuseful sensor
```
Comment (KohliDev) [inline packages/examples/kura/integration_with_kura.py]: Add comment as:
Filter out the sensor not required

File: packages/examples/kura/integration_with_kura.py [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+                    if key_metric == "assetName":
+                        continue
+                    if key_metric.find('stamp') >= 0:
+                        continue
+
+                    if key_metric in Msg_dic.keys(): #exist update msg
+
+                        Msg_dic[key_metric]['msg'] = Jsonpayload[key_metric]
+                        Msg_dic[key_metric]['flag'] = 0
+
```
Comment (KohliDev) [inline packages/examples/kura/integration_with_kura.py]: extra line spacing remove

File: packages/examples/kura/integration_with_kura.py [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+                            userdata['self'].metrics.append(reg_metric)
+                            Msg_dic[key_metric]['metric'] = reg_metric
+
+                        except Exception as e:
+                            log.error(
+                                'Exception while loading metric {0} for Edge System {1}'.format
+                                (key_metric, str(e)))
+
+
+            else: #new device enrolling
```
Comment (KohliDev) [inline packages/examples/kura/integration_with_kura.py]: Add the comment above else loop:
"New Device enrollment"

File: packages/examples/kura/integration_with_kura.py [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+                    else:
+
+                        Msg_dic[key_metric] = {}
+                        Msg_dic[key_metric]['msg'] = Jsonpayload[key_metric]
+                        Msg_dic[key_metric]['flag'] = 1
+
+                        exec(callbackFunc_str % (key_metric, key_metric)) # Dynamically generating callback function
+
+                        try:
+                            metric_ins = Metric(name=key_metric,
```
Comment (KohliDev) [inline packages/examples/kura/integration_with_kura.py]: We need to discuss this approach for every metric are you creating the collecting cycle or will it be done only once?

File: packages/examples/kura/integration_with_kura.py [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+                    raise
+
+                Gatewaydict[Jsonpayload['assetName']] = reg_device
+        else:
+            log.info("Jsonpayload empty")
+    else:
+        log.info("another topic")
+            
+class PackageClass(LiotaPackage):
+
```
Comment (KohliDev) [inline packages/examples/kura/integration_with_kura.py]: Add comment connecting to local MQTT Broker using "MqttDeviceComms"

File: packages/examples/kura/README.txt [hunk 13]
```diff
@@ -0,0 +1,15 @@
+Integrate Kura into Liota
+
+Kura aims at offering a Java/OSGi-based container for M2M applications running in service gateways. Kura provides or, when available, aggregates open source implementations for the most common services needed by M2M applications. Kura components are designed as configurable OSGi Declarative Service exposing service API and raising events. While several Kura components are in pure Java, others are invoked through JNI and have a dependency on the Linux operating system.So it’s worth explore how to integrate this into Liota. 
+
+This directory includes:
+File 1: integration_with_kura.py
+
+
+How to use it:
+
```
Comment (KohliDev) [inline packages/examples/kura/README.txt]: Add link/step on how to configure Kura on IoT Gateway.

File: packages/examples/kura/README.txt [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+1.sha1sum integration_with_kura.py --> 16013ef7a95f59537bc5ab1d91a5c599c9cb8ab4
+2.liotapkg.sh load integration_with_kura 16013ef7a95f59537bc5ab1d91a5c599c9cb8ab4
+3.config Kura UI to publish data
+
+the integration_with_kura.py will auto add sensor.
```
Comment (KohliDev) [inline packages/examples/kura/README.txt]: rename "integration_with_kura.py" as "kura_integrate_iotcc.py"

File: packages/examples/kura/integration_with_kura.py [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+# Lambda Function Multiplier uses the above settings for calculating retry and delay logic
+lfm = lambda x: x * no_of_edge_system_in_thousands
+retry_attempts = lfm(no_of_retries_for_connection)
+delay_retries = random.randint(lfm(delay_retries_min), lfm(delay_retries_max))
+
+Msg_dic = {}                #{"demo": {"flag": 0, "msg": 0, "metric": metric}}
+Gatewaydict = {}            #{"assetName": reg_device}
+
+callbackFunc_str = '''def metric_callback_%s():    return Msg_dic['%s']['msg']''' # update the data by use Msg_dic disc
```
Comment (KohliDev) [inline packages/examples/kura/integration_with_kura.py]: move the comment above and add more detail abourt callbackFunc_str

File: packages/examples/kura/integration_with_kura.py [hunk 16]
```diff
@@ -0,0 +1,256 @@
...
+
+"""Subscibe callback function"""
+def sub_callback(self, client, userdata, msg):
+
+    if (msg.topic.split("/").__len__() == 4) and (msg.topic.split("/")[3] == "pulse-topic"): # msg.topic = "pulse-account-name/pulseid/pulse/pulse-topic"
+
+        Jsonpayload = json.loads(msg.payload)['metrics']
```
Comment (KohliDev) [inline packages/examples/kura/integration_with_kura.py]: use python naming convention everywhere 
json_payload

**Review by KohliDev — CHANGES_REQUESTED** [review]

File: packages/examples/kura/integration_with_kura.py [hunk 17]
```diff
@@ -0,0 +1,256 @@
...
+delay_retries_min = 600
+# Retry delay Max Value in seconds
+delay_retries_max = 1800
+
+# Lambda Function Multiplier uses the above settings for calculating retry and delay logic
+lfm = lambda x: x * no_of_edge_system_in_thousands
+retry_attempts = lfm(no_of_retries_for_connection)
+delay_retries = random.randint(lfm(delay_retries_min), lfm(delay_retries_max))
+
+Msg_dic = {}                #{"demo": {"flag": 0, "msg": 0, "metric": metric}}
```
Comment (KohliDev) [inline packages/examples/kura/integration_with_kura.py]: use python naming convention:

msg_dict
gateway_dict

File: packages/examples/kura/README.txt [hunk 18]
```diff
@@ -0,0 +1,20 @@
+Integrate Kura into Liota
+
+Kura aims at offering a Java/OSGi-based container for M2M applications running in service gateways. Kura provides or, when available, aggregates open source implementations for the most common services needed by M2M applications. Kura components are designed as configurable OSGi Declarative Service exposing service API and raising events. While several Kura components are in pure Java, others are invoked through JNI and have a dependency on the Linux operating system.So it’s worth explore how to integrate this into Liota. 
+
+This directory includes:
+File 1: integration_with_kura.py
```
Comment (KohliDev) [inline packages/examples/kura/README.txt]: Please mention the correct filename

File: packages/examples/kura/README.txt [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+
+
+How to use it:
+
+1.sha1sum integration_with_kura.py --> 16013ef7a95f59537bc5ab1d91a5c599c9cb8ab4
```
Comment (KohliDev) [inline packages/examples/kura/README.txt]: Please mention the correct filename

File: packages/examples/kura/README.txt (hunk 13 above)
Comment (KohliDev) [reply packages/examples/kura/README.txt]: Please add the steps to setup Kura on IoT Gateway

File: packages/examples/kura/kura_integrate_iotcc.py [hunk 20]
```diff
@@ -0,0 +1,258 @@
...
+                device_dict[json_payload['assetName']] = reg_device
+        else:
+            log.info("json_payload empty")
+    else:
+        log.info("another topic")
+
+"""connecting to local MQTT Broker using "MqttDeviceComms"""
+class PackageClass(LiotaPackage):
+
+    BrokerIP = "127.0.0.1"
```
Comment (KohliDev) [inline packages/examples/kura/kura_integrate_iotcc.py]: rename variables:
broker_ip
broker_port
broker_topic

File: packages/examples/kura/kura_integrate_iotcc.py [hunk 21]
```diff
@@ -0,0 +1,258 @@
...
+# Lambda Function Multiplier uses the above settings for calculating retry and delay logic
+lfm = lambda x: x * no_of_edge_system_in_thousands
+retry_attempts = lfm(no_of_retries_for_connection)
+delay_retries = random.randint(lfm(delay_retries_min), lfm(delay_retries_max))
+
+msg_dict = {}               #{"demo": {"flag": 0, "msg": 0, "metric": metric}}
+device_dict = {}            #{"assetName": reg_device}
+
+# Create the callback function dynamically  , this function used to update the data by use msg_dict disc
+callbackFunc_str = '''def metric_callback_%s():    return msg_dict['%s']['msg']'''
```
Comment (KohliDev) [inline packages/examples/kura/kura_integrate_iotcc.py]: rename:
callback_func_str

**Review by KohliDev — APPROVED** [review]
Some minor comments to fix.

**Review by KohliDev — APPROVED** [review]

**Pull request merged**